ramips: add support for ZyXEL Keenetic Giga III

remittor opened the pull request:
**ZyXEL Keenetic Giga III** is a 2.4/5 GHz band 11ac router, based on MT7621ST.

Specification:
-------------
* SoC: MediaTek MT7621ST (1C/2T)
* RAM: 256 MB DDR3
* Flash: 128 MB NAND S34ML01G200TFI00
* Ethernet: 5x 10/100/1000 Mbps
* WIFI: 5 GHz   MT7612EN nac
* WIFI: 2.4 GHz MT7602EN bgn
* USB: 1x 2.0, 1x 3.0
* BTN: Power, Reset, WPS, FN1, FN2
* LEDS: Power(Green),Fn(Green),Wan(Green),WiFi2(Green),WiFi5(Green),USB1(Green),USB2(Green)
* UART: present as five pads without through-holes on the PCB.
Pads are located above 2.4G LED and under board logotype (GND/empty/RX/TX/Vcc).
UART uses 3.3V and settings: 57600-8-N-1

MAC addresses as verified by stock firmware:
----------------------------------------------

| Interface   |       MAC         | Factory | Format |
|-------------|-------------------|---------|--------|
| WiFi (2.4g) | xx:xx:xx:xx:xx:20 | 0x4     | binary |
| LAN         | xx:xx:xx:xx:xx:20 | 0x4     | binary |
| WAN (label) | xx:xx:xx:xx:xx:21 | 0x28    | binary |
| WiFi (5g)   | xx:xx:xx:xx:xx:22 | 0x8004  | binary |

Installation via U-Boot:
------------------------
* Download the latest OpenWrt firmware factory-image and rename it to kgiga3_recovery.bin
* Set up a Tftp server on a PC (e.g. Tftpd32) and place the firmware image to the root directory of the server.
* Power off the router and use a twisted pair cable to connect the PC to any of the router's LAN ports.
* Configure the network adapter of the PC to use IP address 192.168.1.2 and subnet mask 255.255.255.0.
* Power up the router while holding the reset button pressed.
* Wait approximately for 5 seconds and then release the reset button.
* The router should download the firmware via TFTP and complete flashing in a few minutes.
* Connect with SSH to 192.168.1.1 and set a root password or browse to http://192.168.1.1 if LuCI is installed.

Installation via Keenetic-OS GUI:
------------------------------------------------
* Set the IP address 192.168.1.1 for the connected device (in the KeeneticOS settings and in the network connection properties).
* Open the page "http://192.168.1.1/a" in the browser, enter the command "more proc:/dual_image/boot_active" (without quotes) in the "Command" field and click the "Send request" button.
* A response will appear in the browser, from which you need to select only the value of the "message" parameter.
* If the value of the "message" parameter is not equal to 2, then you need to force flash any official firmware and go to the first step of the instruction.
* To install OpenWRT, go to the "General Settings" page and in the "System Files" section, click on "firmware".
* Then click the "Replace File" button.
* In the window that appears, specify the location of the file "openwrt-ramips-mt7621-zyxel_keenetic-giga-iii-squashfs-factory.bin".
* The installation of OpenWRT on the device will begin.
* After installing the firmware, the device will reboot itself and start OpenWRT (takes about 60...90 seconds).

Revert to the stock firmware via OpenWrt GUI:
------------------------------------------------
* Upload the stock image using "Flash Firmware" interface.
* Select "Force update" and continuing upgrade.

jospezial commented:
Telnet? SSH would be better.

remittor commented:
> Telnet? SSH would be better.

See: https://openwrt.org/toh/zyxel/zyxel_keenetic_omni_rev._a#oem_easy_installation
And many other manuals.

jospezial commented:
The page seems to be not up to date.
It looks as telnet(d) is not in the openwrt images.
https://github.com/openwrt/openwrt/commit/d196b1fc2eddf7cbab4351452b6fc9b41f15b142

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+#include "mt7621.dtsi"
```
Please add an SPDX license identifier.

remittor replied:
fixed

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+	chosen {
+		bootargs = "console=ttyS0,57600n8";
+	};
```
This block can be dropped, it's in the DTSI.

remittor replied:
fixed

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
```
Use gpio-keys instead. Just look at any other DTS file for mt7621.

remittor replied:
fixed

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+			label = "reset";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			/* debounce-interval = <60>; */
```
Remove that comment.

remittor replied:
fixed

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+			label = "fn2";
+			gpios = <&gpio 10 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_1>;
+		};
```
These blocks should be sorted by something. Either GPIO number or name.

remittor replied:
fixed

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+			gpios = <&gpio 12 GPIO_ACTIVE_HIGH>;
+		};
+
+		internet { /* WAN */
```
Remove the comment.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+			partition@0 {
+				label = "ubi";         // special name for creating UBIFS
+				reg = <0x0 0x7500000>; // 117 MB
```
Remove comments

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+};
+
+&nand {
+	status = "okay";
```
Empty line after status.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x80000>; // 512 KB
```
Again, remove comments. Also all below, I won't comment again ...

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+		/* U-Boot expects to find kernel image at 0x180000 */
+		partition@180000 {
+			label = "firmware";         // Firmware_1
+			reg = <0x180000 0x3E40000>; // 122 MB
```
Please use lower case for hex letters.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+		ubiconcat1: partition@4140000 {
+			label = "ubiconcat1";        // Firmware_2
+			reg = <0x4140000 0x3AC0000>; // 58 MB
```
lower case.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+		/* U-Boot used latest 32 BLOCK for factory bbt table */
+		partition@7C00000 { // 128MiB - 32*128KiB = 0x7C00000
+			label = "mtk-nand-bbt";
+			reg = <0x7C00000 0x400000>; // 4 MB
```
lower case.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+		};
+
+		/* U-Boot used latest 32 BLOCK for factory bbt table */
+		partition@7C00000 { // 128MiB - 32*128KiB = 0x7C00000
```
lower case.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+&xhci {
+	status = "okay";
+};
```
This can be dropped, xhci is enabled by default.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x8000>;
+		nvmem-cells = <&macaddr_factory_8004>;
+		nvmem-cell-names = "mac-address";
```
nvmem-cells can be dropped here, this location is loaded automatically from mtd-eeprom.

https://openwrt.org/docs/guide-developer/mac.address#mac_address_pulled_by_driver

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+};
+
+&pcie0 { // MT7612EN wifi5g
+	wifi0: mt76@0,0 {  
```
wifi0: and wifi1: DT labels are unused and can be dropped.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+		nvmem-cells = <&macaddr_factory_4>;
+		nvmem-cell-names = "mac-address";
```
Like above, nvmem-cell* can be dropped.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+&gmac0 {
+	nvmem-cells = <&macaddr_factory_28>;
+	nvmem-cell-names = "mac-address";
```
This does not match your description in commit message.

While it's uncommon anyway (typically LAN is 0x28 and WAN is 0x2e), based on your description I'd expect 0x4 here.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+	macaddr_factory_8004: macaddr@8004 { /* wifi5 mac (wan + 1) */
+		reg = <0x8004 0x6>;
+	};
```
This last block can be dropped, since we get it from mtd-eeprom.

remittor replied:
Although not currently used, it should be kept for information.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+
+&state_default {
+	gpio {
+		groups = "uart3", "jtag", "wdt";
```
On a quick look, uart2 seems to be missing.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
+define Device/zyxel_keenetic-giga-iii
+  $(Device/dsa-migration)
+  KERNEL_SIZE := 4096k  # kernel partition size
+  IMAGE_SIZE := 50m
```
Please specify in "k" instead, thats common practice.

remittor replied:
Megabytes are also used:
```
$ grep -r --include "*.mk" -E "IMAGE_SIZE := [0-9]{1,3}m"
layerscape/image/armv8_64b.mk:  IMAGE_SIZE := 64m
layerscape/image/armv7.mk:  IMAGE_SIZE := 64m
ath79/image/nand.mk:  IMAGE_SIZE := 116m
```
And this suffix is handled correctly: https://github.com/openwrt/openwrt/blob/ac03e246351fbfa98e98f0319cce3a8663f836d3/include/image-commands.mk#L171

adschm replied:
Out of multiple hundred uses of this variable, there is indeed a one-digit number of cases, on other targets, that does not use "k". That's essentially what common practice means.
So, please change it after you have proven my point now.

remittor replied:
OK. So I'll write the value 51000k

adschm replied:
You have no clue what this value does, do you?

remittor replied:
Do you want me to explain to you what this code does?
https://github.com/openwrt/openwrt/blob/ac03e246351fbfa98e98f0319cce3a8663f836d3/include/image-commands.mk#L169-L176
I have not found the use of this variable anywhere else.

adschm replied:
"check-size" checks the size of the final image. For devices with small flash, typically 4M/8M, you will quickly manage to create images that are bigger than the available amount of firmware partition space to write them to. Since that might result in a broken or even unbootable device, the build routine checks the image size and will abort the build process if the "image" is too large when reaching the check-size check.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
+  PAGESIZE := 2048      # NAND page size
+  DEVICE_VENDOR := ZyXEL
+  DEVICE_MODEL := Keenetic Giga III
+  SUPPORTED_DEVICES += kng_re
```
This line should be dropped, unless it's needed for factory flashing.

remittor replied:
This variable `PAGESIZE` must be specified as well as the variable `UBINIZE_OPTS := -E 5`.
See: https://github.com/openwrt/openwrt/blob/ac03e246351fbfa98e98f0319cce3a8663f836d3/include/image-commands.mk#L101-L114
See ubinize: https://github.com/sigma-star/mtd-utils/blob/f0b32c7ba5136151a961c3a43c8ff430b33d7c33/ubi-utils/ubinize.c#L47-L48
If we use a method `append-ubi`, then we must specify the variable `PAGESIZE`.

adschm replied:
Funny, how you discuss all of the variables here except the line I was referring to.
I was referring to the line directly before my comment, which is SUPPORTED_DEVICES ...

remittor replied:
You will be surprised when you enter the command:
`grep -r --include "*.mk" -n -E "SUPPORTED_DEVICES +" | wc -l`
I doubt very much that all found devices need to specify this value for factory flashing.

remittor replied:
grep finds several Keenetic:
```
ramips/image/mt7620.mk:1329:  SUPPORTED_DEVICES += kn_rc
ramips/image/mt7620.mk:1342:  SUPPORTED_DEVICES += kn_rf
ramips/image/mt7620.mk:1356:  SUPPORTED_DEVICES += kng_rc
```

adschm replied:
Have you ever considered the possibility that I know what I'm speaking of and instead of you might actually know what SUPPORTED_DEVICES _is_ good for?

remittor replied:
Я прежде всего доверяю написанному коду и скриптам.
The «kng_re» value is stored in the bootloader configs ("U-Config" partition).
```
servicehost=ndss.11.zyxel.ndmsystems.com
servicepass=6Wg_____480
sernumb=S150_____22
ndmhwid=kng_re
servicetag=500______776
wlanssid=Keenetic-8204
wlankey=S3i___HZ
wlanwps=658___88
country=RU
```

adschm replied:
Interesting. Now, would you please remove the SUPPORTED_DEVICES line?

SUPPORTED_DEVICES is for upgrades between different OpenWrt firmwares only and since this is the first OpenWrt firmware (unless OEM firmware is, which is why I asked the initial question), we don't need a custom second value here.

remittor replied:
OK. I will delete it. Looking at the code, I realized that it has a double meaning (not only about what you said).

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
+  DEVICE_MODEL := Keenetic Giga III
+  SUPPORTED_DEVICES += kng_re
+  DEVICE_PACKAGES := kmod-mt76x2 kmod-usb2 kmod-usb3
+  DEVICE_PACKAGES += kmod-usb-ledtrig-usbport kmod-usb-storage
```
Please use a multiline assignment for one variable like all the others.
I don't think usb-storage should be included by default.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
+  VER_NUM := $$(shell cut -d '-' -f 1 <<< "$$(VERSION_NUMBER)" )
+  VER_DATE := $$(shell date --utc -d @$$(SOURCE_DATE_EPOCH) +%y%m%d)
+  VER_FULL := $$(VER_NUM) $$(VER_DATE)
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | uImage none -n 'KN-GIGA3 $$(VERSION_DIST) $$(VER_FULL)'
```
please create a Build recipe for that instead of introducing local variables here.

remittor replied:
I think this is completely superfluous.
Did you talk about something similar?
```
define Device/repo-version
  REPO_VER := $$(shell cut -d '-' -f 1 <<< "$$(VERSION_NUMBER)" )
  REPO_DATE := $$(shell date --utc -d @$$(SOURCE_DATE_EPOCH) +%y%m%d)
  REPO_VER_DATE := $$(REPO_VER) $$(REPO_DATE)
  REPO_FULL_VER := $$(VERSION_DIST) $$(REPO_VER) $$(REPO_DATE)
endif
```

adschm replied:
No, this is a common device definition. I'm talking about a build recipe, these are the blocks starting with "Build/", i.e. those you are using when you put all these nice terms like "kernel-bin" or "loader-kernel" ...

adschm replied:
Some thing like this (just typed it without too much care):

```
define Build/my_special_uimage_command
  $(eval repo_ver=$$(shell cut -d '-' -f 1 <<< "$$(VERSION_NUMBER)"))
  $(eval repo_date=$$(shell date --utc -d @$$(SOURCE_DATE_EPOCH) +%y%m%d))
  uImage none -n '$(word 1, $(1)) $$(VERSION_DIST) $(repo_ver) $(repo_date)'
endef

[...]
KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | my_special_uimage_command KN-GIGA3
[...]
```

remittor replied:
Maybe you can then do what you like with a separate commit?
I really think that making a specially separate recipe for one device is a complete overkill.

remittor replied:
And in which file should this recipe be placed?
Is it possible in `target/linux/ramips/image/Makefile` ?

adschm replied:
This depends on the devices it will be used for. If it's just for one device, you can put in the same file as the device definition (of course, at the top and keeping proper sorting of recipes).

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
+  UBINIZE_OPTS := -E 5   # EOD marks to "hide" factory sig at EOF
+  IMAGES := sysupgrade.tar
+  IMAGE/sysupgrade.tar := sysupgrade-tar | append-metadata
+  IMAGES += factory.bin
```
IMAGES in one line.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1685,6 +1685,31 @@ define Device/zio_freezio
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi
+  IMAGE/factory.bin += | check-size | pad-to $$(BLOCKSIZE)
+  IMAGE/factory.bin += | zyimage -d 10368 -v "$$(DEVICE_VENDOR) $$(DEVICE_MODEL) $$(VERSION_DIST) $$(VER_FULL)"
```
One multiline command. As above, please create a Build recipe for the last step.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -122,6 +122,11 @@ xiaomi,redmi-router-ac2100)
+	ucidef_set_led_netdev  "internet" "internet"  "green:internet" "wan"
+	ucidef_set_led_netdev  "wifi5g"   "Wifi 5G"   "green:wifi5G"   "wlan0"
+	ucidef_set_led_netdev  "wifi2g"   "Wifi 2.4G" "green:wifi2G"   "wlan1"
+	;;
```
Please use the same formatting as anybody else.

Wifi setup can be dropped here since you have DT triggers.

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+	model = "ZyXEL Keenetic Giga III";
+
+	aliases {
+		label-mac-device = &gmac0;
```
This should be &wan (you should add the DT label as described below).

adschm commented on target/linux/ramips/dts/mt7621_zyxel_keenetic-giga-iii.dts:
```diff
@@ -0,0 +1,301 @@
+			label = "lan1";
+		};
+
+		port@4 {
```
Use "wan: port@4 {" here, so you can use the mac address for label-mac-device.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -157,6 +157,11 @@ ramips_setup_macs()
+		lan_mac=$(mtd_get_mac_binary factory 0x4)
+		wan_mac=$(mtd_get_mac_binary factory 0x28)
+		label_mac=$wan_mac
+		;;
```
This can be dropped entirely when you've fixed your DTS.

adschm commented:
I'm sorry for being a bit rude here. But you obviously try to teach me on things that you don't really understand yourself, and that's not helping. Note that I have commit access to this repository, so you can except that when I give you advice on how to fix things, there is a high chance that my pointer is correct.

There is a lot of submissions to review and I don't have time for arguments like that, so please, again, excuse my direct address here.

remittor commented:
> I'm sorry for being a bit rude here.

It's OK. You have found many bugs.

remittor commented:
ping @adschm

nm17 commented:
@remittor Any plans on continuing the development?

remittor commented:
@nm17, So it's all done. You can merge and build the firmware.

jospezial commented:
You did not exchange the telnet advise with ssh.
https://openwrt.org/toh/zyxel/zyxel_keenetic_omni_rev._a#oem_easy_installation should get an update to ssh.

nm17 commented:
@remittor why is the pr not passing the review?

neheb commented:
Must be rebased against current master.

remittor commented:
> Must be rebased against current master.

Done

remittor commented:
> Two changes

Why two changes?

> GPIO switch LED support. The Netgear wax202 PR has an example.

Should I add to the `01_leds` file using the `ucidef_set_led_usbport`?
No device based on mt7621 uses this specifier.

neheb commented:
Then no.

I'm talking about switch LEDs, not normal GPIO ones.

arinc9 commented:
> > Implement the changes in #10238
> 
> Added this as a separate commit. But I haven't tested these changes.

The commit log is too vague to be accepted. The main commit isn't merged yet anyway, so just squash em.

remittor committed:
ramips: add support for ZyXEL Keenetic Giga III

ZyXEL Keenetic Giga III is a 2.4/5 GHz band 11ac router

Specification:
* SoC: MediaTek MT7621ST (1C/2T)
* RAM: 256 MB DDR3
* Flash: 128 MB NAND S34ML01G200TFI00
* Ethernet: 5x 10/100/1000 Mbps
* WIFI: 5 GHz   MT7612EN nac
* WIFI: 2.4 GHz MT7602EN bgn
* USB: 1x 2.0, 1x 3.0
* BTN: Power, Reset, WPS, FN1, FN2
* LEDS: Power(Green),Fn(Green),Wan(Green),WiFi2(Green),WiFi5(Green),
  USB1(Green),USB2(Green)
* UART: present as five pads without through-holes on the PCB.
  Pads are located above 2.4G LED and under board logotype
  (GND/empty/RX/TX/Vcc).
  UART uses 3.3V and settings: 57600-8-N-1

MAC addresses as verified by stock firmware:

| Interface   |       MAC         | Factory | Format |
|-------------+-------------------+---------+--------|
| WiFi (2.4g) | xx:xx:xx:xx:xx:20 | 0x4     | binary |
| LAN         | xx:xx:xx:xx:xx:20 | 0x4     | binary |
| WAN (label) | xx:xx:xx:xx:xx:21 | 0x28    | binary |
| WiFi (5g)   | xx:xx:xx:xx:xx:22 | 0x8004  | binary |

Installation via U-Boot:
* Download the latest OpenWrt firmware factory-image and rename it to
  kgiga3_recovery.bin
* Set up a Tftp server on a PC (e.g. Tftpd32) and place the firmware
  image to the root directory of the server.
* Power off the router and use a twisted pair cable to connect the PC
  to any of the router's LAN ports.
* Configure the network adapter of the PC to use IP address 192.168.1.2
  and subnet mask 255.255.255.0.
* Power up the router while holding the reset button pressed.
* Wait approximately for 5 seconds and then release the reset button.
* The router should download the firmware via TFTP and complete flashing
  in a few minutes.
* Connect with SSH to 192.168.1.1 and set a root password or browse to
  http://192.168.1.1 if LuCI is installed.

Installation via Keenetic-OS GUI:
* Set the IP address 192.168.1.1 for the connected device (in the
  KeeneticOS settings and in the network connection properties).
* Open the page "http://192.168.1.1/a" in the browser, enter the
  command "more proc:/dual_image/boot_active" (without quotes) in
  the "Command" field and click the "Send request" button.
* A response will appear in the browser, from which you need to select
  only the value of the "message" parameter.
* If the value of the "message" parameter is not equal to 2, then you
  need to force flash any official firmware (Keenetic-OS) and go to
  the first step of the instruction.
* To install OpenWRT, go to the "General Settings" page and in the
  "System Files" section, click on "firmware".
* Then click the "Replace File" button.
* In the window that appears, specify the location of the file
  "openwrt-ramips-mt7621-zyxel_keenetic-giga-iii-squashfs-factory.bin".
* The installation of OpenWRT on the device will begin and complete
  flashing in a few minutes.
* Browse to http://192.168.1.1 and set root password.

Revert to the stock firmware via OpenWrt GUI:
* Upload the stock image using "Flash Firmware" interface.
* Select "Force update" and continuing upgrade.

Signed-off-by: Oleg S <remittor@gmail.com>

remittor commented:
> Implement the changes in #10238

Added this patch. I also tested the work with these changes (without testing the maximum throughput).

KashBrus commented:
Hi @remittor any chance or time gap when it would be merged?
Will it be available from official builds?

hauke commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -120,6 +120,21 @@ platform_do_upgrade() {
+			reboot -f
+			exit 1
+		fi
+		CI_KERNPART="kernel"
```
`CI_KERNPART="kernel"` is already the default. No need to set it here.

hauke commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -120,6 +120,21 @@ platform_do_upgrade() {
+	zyxel,keenetic-giga-iii)
+		local magic=$( nand_get_magic_long "$1" 0 )
+		echo "Firmware image $1 contain magic $magic"
+		if [ "$magic" == "27051956" ]; then
```
Why is this special handling needed here?
Please add a comment.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -2274,6 +2274,24 @@ define Device/zio_freezio
+  IMAGE_SIZE := 51000k
+  BLOCKSIZE := 128k     # NAND erase size
+  PAGESIZE := 2048      # NAND page size
+  UBINIZE_OPTS := -E 5  # EOD marks to "hide" factory sig at EOF
```
Please replace these nand common definitions with `$(Device/nand)`.
ref: https://github.com/openwrt/openwrt/commit/b09a83808a3a4f6ff2ff0cf1f12fdc4193357896

Nekotekina commented:
Why did I have such a bad luck choosing a router that isn't officially supported...
Nevertheless I installed openwrt with this PR patch and it seems working, more or less. Thank you.

GoetzGoerisch commented:
Thank you for your PR.

Just a heads up, we had a recent [treewide change from ZyXEL to Zyxel](https://github.com/openwrt/openwrt/commit/3774f3272ee976b922306a881b6eb603e79f81ed).
Please adjust your commit message and code accordingly. Thank you!
